### ISSUE TITLE
Bad write retry error with python 2.7.10

### ISSUE THREAD

**Issue by Nagriar** [algolia/algoliasearch-client-python#30]
When I try to execute
```python
def test_network(self):
        batch = []
        for i in range(1, 3000):
            batch.append({'action': 'addObject', 'body': {
                'test1': 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'
            }})
        self.index.batch(batch)
```

I receive the error: AlgoliaException: Unreachable hosts: {'YLGNV73XLW-2.algolianet.com': "[('SSL routines', 'ssl3_write_pending', 'bad write retry')]", 'YLGNV73XLW.algolia.net': "[('SSL routines', 'ssl3_write_pending', 'bad write retry')]", 'YLGNV73XLW-3.algolianet.com': "[('SSL routines', 'ssl3_write_pending', 'bad write retry')]", 'YLGNV73XLW-1.algolianet.com': "[('SSL routines', 'ssl3_write_pending', 'bad write retry')]"}

If I use python 3.4, the issue is fixed, it's maybe related to https://hg.python.org/cpython/rev/60310223d075/

**Comment by dethi** [comment]
Interesting issue :open_mouth: 
For easier reading, the same snippet formatted:
```python
def test_network(self):
    batch = []
    for i in range (3000):
        batch.append({'action': 'addObject', 'body': {'test1': 'aaaaaaaaaaaaaaaaaaaaaaaaaaa'}})
    self.index.batch(batch)
```

**Comment by dethi** [comment]
Okey, so the issue comes from PyOpenSSL and Python 2.7

**Comment by redox** [comment]
> The issue comes from PyOpenSSL and Python 2.7

:dancer:

**Comment by dethi** [comment]
We use PyOpenSSL on Python 2.X for the SNI Support... :(

**Comment by Nagriar** [comment]
Why 2.6 is working on travis? it's a regression?

**Comment by dethi** [comment]
https://docs.python.org/2/library/ssl.html#ssl.HAS_SNI
Ok so, PyOpenSSL is only needed for Python 2.6 because Python 2.7.9+ comes with SNI Support :)

**Comment by redox** [comment]
> So PyOpenSSL is only needed for Python 2.6 because Python 2.7.9+ comes with SNI Support :)

Interesting :) So do you think we could depend on it based on the version of python?

**Comment by dethi** [comment]
@redox Well, I tried last month to use "package_name;python_version >= 'X.X' but without success. **pip** has some issue with this syntax and many OS are shipped with old version of **pip**, which is bad... 

@Nagriar yes I think, but I can't find any opened issue about it.

**Comment by Nagriar** [comment]
The fix is on the master branch of urllib3 (https://github.com/shazow/urllib3/blob/master/CHANGES.rst#dev-master). We just need to wait the release to fix this issue.

**Comment by dethi** [comment]
Cool :)

**Comment by Nagriar** [comment]
:boom: Version 1.13 released

**Comment by dustincoates** [comment]
@Nagriar @dethi Is there a workaround for those who can't upgrade urllib3 because of dependency issues elsewhere?

**Comment by dethi** [comment]
@dustincoates if you can uninstall PyOpenSSL, it should work.

**Comment by dustincoates** [comment]
Thanks @dethi. I'll see if that's an option for them.

**Comment by redox** [comment]
Just to make it clear for people reading that issue later on:
 - the issue is indeed in `urllib3` and has been fixed with https://github.com/shazow/urllib3/pull/719
 - BUT upgrading `urllib3` **won't solve** the issue because `requests` embed it's own version of `urllib3`; so the API client won't use that updated version
 - the good thing is that `requests` upgraded to `urllib3 1.13` in their `2.9.x` release

**tldr;**

```
pip install requests[security]==2.9.1
```
